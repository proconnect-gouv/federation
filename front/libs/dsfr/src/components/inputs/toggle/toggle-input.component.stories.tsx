import { ComponentMeta } from '@storybook/react';

import { ToggleInputComponent, ToggleInputComponentProps } from './toggle-input.component';

export default {
  component: ToggleInputComponent,
  title: 'DSFR/components/toogle/ToggleInputComponent',
} as ComponentMeta<typeof ToggleInputComponent>;

export const Default = (args: ToggleInputComponentProps) => <ToggleInputComponent {...args} />;

Default.args = {
  input: {
    name: 'checkbox',
  },
} as ToggleInputComponentProps;
