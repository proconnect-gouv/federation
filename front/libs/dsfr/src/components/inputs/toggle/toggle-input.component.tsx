import React from 'react';
import { FieldInputProps } from 'react-final-form';

export interface ToggleInputComponentProps {
  disabled?: boolean;
  // @NOTE la regle est desactivée car le type provient de la librairie react-final-form
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  input: FieldInputProps<any, HTMLElement>;
}

export const ToggleInputComponent: React.FC<ToggleInputComponentProps> = React.memo(
  ({ disabled, input }: ToggleInputComponentProps) => (
    <input
      // @NOTE on autorise la destructuration des proprietes car
      // les propprietes fournies par react-final-form sont issues d'un contexte contrôlé
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...input}
      aria-describedby={`${input.name}-hint-text`}
      className="fr-toggle__input"
      data-testid={`field-toggle-input-${input.name}`}
      disabled={disabled}
      id={input.name}
      type="checkbox"
      onChange={input.onChange}
    />
  ),
);

ToggleInputComponent.defaultProps = {
  disabled: false,
};

ToggleInputComponent.displayName = 'ToggleInputComponent';
