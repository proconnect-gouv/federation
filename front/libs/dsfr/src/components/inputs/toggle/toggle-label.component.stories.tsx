import { ComponentMeta } from '@storybook/react';

import { ToggleLabelComponent, ToggleLabelComponentProps } from './toggle-label.component';

export default {
  component: ToggleLabelComponent,
  title: 'DSFR/components/toogle/ToggleLabelComponent',
} as ComponentMeta<typeof ToggleLabelComponent>;

export const Default = (args: ToggleLabelComponentProps) => <ToggleLabelComponent {...args} />;

export const WithFunctionLabel = (args: ToggleLabelComponentProps) => (
  <ToggleLabelComponent {...args} />
);

Default.args = {
  input: {
    name: 'checkbox',
  },
  label: 'Toggle label',
  legend: { checked: 'Activé', unchecked: 'Désactivé' },
} as ToggleLabelComponentProps;

WithFunctionLabel.args = {
  input: {
    name: 'checkbox',
  },
  label: () => 'Result function label',
  legend: { checked: 'Activé', unchecked: 'Désactivé' },
} as unknown as ToggleLabelComponentProps;
