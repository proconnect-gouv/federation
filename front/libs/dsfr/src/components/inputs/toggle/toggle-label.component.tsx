import React from 'react';
import { FieldInputProps } from 'react-final-form';

import { CheckableLegend } from '../../../interfaces';

interface ToggleLabelComponentProps {
  // @NOTE la regle est desactivée car le type provient de la librairie react-final-form
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  input: FieldInputProps<any, HTMLElement>;
  label: string | Function;
  legend?: CheckableLegend;
}

const defaultLegend = { checked: 'Activé', unchecked: 'Désactivé' };

export const ToggleLabelComponent: React.FC<ToggleLabelComponentProps> = React.memo(
  ({ input, label, legend = defaultLegend }: ToggleLabelComponentProps) => {
    const isLabelString = typeof label === 'string';
    const isLabelFunction = typeof label === 'function';
    return (
      <label
        className="fr-toggle__label"
        data-fr-checked-label={legend.checked}
        data-fr-unchecked-label={legend.unchecked}
        htmlFor={input.name}>
        {isLabelString && label}
        {isLabelFunction && label(input.checked)}
      </label>
    );
  },
);

ToggleLabelComponent.defaultProps = {
  legend: undefined,
};

ToggleLabelComponent.displayName = 'ToggleLabelComponent';
