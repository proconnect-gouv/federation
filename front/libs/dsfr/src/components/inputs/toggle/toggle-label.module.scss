/**
///
/// @NOTE these values are JS managed by official DSFR
/// this file purpose is about overriding these values
/// without implementing potiental conflictual React/VanillaJS
///
*/

.label::before {
  --toggle-status-width: 50px;
}
