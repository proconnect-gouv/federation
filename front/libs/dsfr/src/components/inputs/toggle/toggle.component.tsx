import classnames from 'classnames';
import React from 'react';
import type { FieldInputProps } from 'react-final-form';

import type { CheckableLegend } from '../../../interfaces';
import { ToggleInputComponent } from './toggle-input.component';
import { ToggleLabelComponent } from './toggle-label.component';

interface ToggleComponentProps {
  // @SEE https://gouvfr.atlassian.net/wiki/spaces/DB/pages/217251933/Case+cocher+-+Checkbox
  disabled?: boolean;
  hint?: string;
  // @NOTE la regle est desactiv√©e car le type provient de la librairie react-final-form
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  input: FieldInputProps<any, HTMLElement>;
  label: string | Function;
  className?: string;
  onUpdate?: (v: boolean) => void;
  legend: CheckableLegend | undefined;
}

export const ToggleComponent = React.memo(
  ({
    className,
    disabled = false,
    hint,
    input,
    label,
    legend,
    onUpdate = undefined,
  }: ToggleComponentProps) => (
    <div className={classnames('fr-toggle', className)}>
      <ToggleInputComponent disabled={disabled} input={input} onUpdate={onUpdate} />
      <ToggleLabelComponent input={input} label={label} legend={legend} />
      {hint && (
        <p className="fr-hint-text" id={`${input.name}-hint-text`}>
          {hint}
        </p>
      )}
    </div>
  ),
);

ToggleComponent.displayName = 'ToggleComponent';
