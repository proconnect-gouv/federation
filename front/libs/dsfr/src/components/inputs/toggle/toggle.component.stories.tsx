import { ComponentMeta } from '@storybook/react';

import { ToggleComponent, ToggleComponentProps } from './toggle.component';

export default {
  component: ToggleComponent,
  title: 'DSFR/components/toogle/ToggleComponent',
} as ComponentMeta<typeof ToggleComponent>;

export const Default = (args: ToggleComponentProps) => <ToggleComponent {...args} />;

export const WithFunctionLabel = (args: ToggleComponentProps) => <ToggleComponent {...args} />;

Default.args = {
  input: {
    name: 'checkbox',
  },
  label: 'Toggle label',
  legend: { checked: 'Activé', unchecked: 'Désactivé' },
} as ToggleComponentProps;

// @TODO: Create a type for this story args
WithFunctionLabel.args = {
  input: {
    name: 'checkbox',
  },
  label: () => 'Result function label',
  legend: { checked: 'Activé', unchecked: 'Désactivé' },
} as unknown as ToggleComponentProps;
