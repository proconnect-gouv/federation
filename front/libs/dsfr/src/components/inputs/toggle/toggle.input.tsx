import React from 'react';
import { Field } from 'react-final-form';

import type { CheckableLegend } from '../../../interfaces';
import { ToggleComponent } from './toggle.component';

interface ToggleInputProps {
  // @SEE https://gouvfr.atlassian.net/wiki/spaces/DB/pages/368935138/Interrupteur+-+Toggle+switch
  disabled?: boolean;
  hint?: string;
  // @NOTE le initialValue est un argument provenant de la lib react-final-Form
  // son typage de base est `any`, notre wrapper est donc obligé d'hérité de ce type
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  initialValue?: any;
  label: string | Function;
  name: string;
  className?: string;
  legend?: CheckableLegend;
  onUpdate?: (v: boolean) => void;
}

export const ToggleInput = React.memo(
  ({
    className,
    disabled = false,
    hint,
    initialValue,
    label,
    legend,
    name,
    onUpdate = undefined,
  }: ToggleInputProps) => (
    <Field initialValue={initialValue} name={name} type="checkbox">
      {({ input }) => (
        <ToggleComponent
          className={className}
          disabled={disabled}
          hint={hint}
          input={input}
          label={label}
          legend={legend}
          onUpdate={onUpdate}
        />
      )}
    </Field>
  ),
);

ToggleInput.displayName = 'ToggleInput';
