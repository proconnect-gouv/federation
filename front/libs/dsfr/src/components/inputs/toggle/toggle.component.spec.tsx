import { render } from '@testing-library/react';
import { FieldInputProps } from 'react-final-form';

import { CheckableLegend } from '../../../interfaces';
import { ToggleComponent } from './toggle.component';

jest.mock('./toggle-input.component');
jest.mock('./toggle-label.component');

describe('ToggleComponent', () => {
  const fieldInputPropsMock = {
    name: 'any-field-name-mock',
    onBlur: () => {},
    onChange: () => {},
    onFocus: () => {},
    value: 'any-field-value-mock',
  } as FieldInputProps<unknown, HTMLElement>;
  const legendMock = { checked: 'foo', unchecked: 'bar' } as CheckableLegend;

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot with default props', () => {
    // When
    const { container } = render(
      <ToggleComponent input={fieldInputPropsMock} label="foobar" legend={legendMock} />,
    );
    // Then
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should match snapshot with hint provided', () => {
    // Given
    const hintMock = 'hint-mock';
    // When
    const { container } = render(
      <ToggleComponent
        hint={hintMock}
        input={fieldInputPropsMock}
        label="foobar"
        legend={legendMock}
      />,
    );
    // Then
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should apply className if provided', () => {
    // Given
    const classNameMock = 'class-name-mock';
    // When
    const { container } = render(
      <ToggleComponent
        className={classNameMock}
        input={fieldInputPropsMock}
        label="foobar"
        legend={legendMock}
      />,
    );
    // Then
    expect(container.firstChild).toHaveClass(classNameMock);
  });
});
