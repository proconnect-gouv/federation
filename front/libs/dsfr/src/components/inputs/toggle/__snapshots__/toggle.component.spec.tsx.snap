// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleComponent should match snapshot with default props 1`] = `
<div
  class="fr-toggle"
>
  <div>
    ToggleInputComponent
  </div>
  <div>
    ToggleLabelComponent
  </div>
</div>
`;

exports[`ToggleComponent should match snapshot with hint provided 1`] = `
<div
  class="fr-toggle"
>
  <div>
    ToggleInputComponent
  </div>
  <div>
    ToggleLabelComponent
  </div>
  <p
    class="fr-hint-text"
    id="any-field-name-mock-hint-text"
  >
    hint-mock
  </p>
</div>
`;
